### DISCUSSION TITLE
Using the Held-Karp TSP lower bound alone

### DISCUSSION THREAD

**Discussion by TryerGit** [google/or-tools#3901] [Routing (and legacy CP) questions]
Hello,
I am interested in using the Held-Karp TSP lower bound for a problem. I found the following page where it indicates that it is implemented in or-tools:

https://developers.google.com/optimization/reference/graph/one_tree_lower_bound/

However, it is not clear to me if to obtain this functionality and use it within my code, I would have to install the entirety of or-tools as specified on this page: https://developers.google.com/optimization/install/cpp/source_windows or this page: https://developers.google.com/optimization/install/cpp/binary_windows

or if I should only use the two headers specified in developers.google.com webpage linked to above: 

https://developers.google.com/optimization/reference/graph/one_tree_lower_bound/HeldWolfeCrowderEvaluator and
https://developers.google.com/optimization/reference/graph/one_tree_lower_bound/VolgenantJonkerEvaluator

Any help is appreciated.
Thank you.

**Comment by sschnug** [comment]
I'm afraid you will need much much more than just those headers.

The bazel-based build files [hint this](https://github.com/google/or-tools/blob/44e95f51b0c84b6c25414f81c0175703053ddfff/ortools/graph/BUILD.bazel#L176):

- There are lots of dependencies including abseil, a huge third party lib (logging, containers, ...).

My personal recommendation would be using one of the two supported build systems and build from sources: either [cmake](https://github.com/google/or-tools/tree/v9.7#build-from-source-using-cmake) or [bazel](https://github.com/google/or-tools/tree/v9.7#build-from-source-using-bazel).

**Comment by TryerGit** [comment]
Thank you. On trying to install via cmake following instructions here: https://developers.google.com/optimization/install/cpp/source_windows

running the command: cmake --build build --config Release --target INSTALL -v

gives:

"C:\Users\Tryer\Downloads\or-tools-stable\or-tools-stable\build\INSTALL.vcxproj" (default target) (1) ->
"C:\Users\Tryer\Downloads\or-tools-stable\or-tools-stable\build\ALL_BUILD.vcxproj" (default target) (2) ->
"C:\Users\Tryer\Downloads\or-tools-stable\or-tools-stable\build\ortools\linear_solver\samples\assignment_groups_mip.vcxproj" (default target) (13) ->
"C:\Users\Tryer\Downloads\or-tools-stable\or-tools-stable\build\ortools.vcxproj" (default target) (86) ->
"C:\Users\Tryer\Downloads\or-tools-stable\or-tools-stable\build\ortools\glop\ortools_glop.vcxproj" (default target) (105) ->
(ClCompile target) ->
  C:\Users\Tryer\Downloads\or-tools-stable\or-tools-stable\build\_deps\absl-src\absl\container\internal\compressed_tuple.h(75): fatal error C1001: Internal compiler error. [C:\Users\Tryer\Downloads\or-tools-stable\or-tools-stable\build\
ortools\glop\ortools_glop.vcxproj]

I am running on Visual Studio 2019.

I will now try a make based install and see how it goes.

**Reply by sschnug** [reply]
[make is deprecated](https://github.com/google/or-tools/issues/3558)

**Comment by TryerGit** [comment]
I tried to follow the install steps from https://developers.google.com/optimization/install/cpp/source_windows this time on a machine with Visual Studio 2022

The last step, of cmake --build build --config Release --target INSTALL -v terminates with the following errors:

```
"C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL.vcxproj" (default target) (1) ->
"C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\ALL_BUILD.vcxproj" (default target) (2) ->
"C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\ortools\linear_solver\samples\assignment_groups_mip.vcxproj" (default target) (13) ->
"C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\ortools.vcxproj" (default target) (86) ->
"C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\ortools\glop\ortools_glop.vcxproj" (default target) (105) ->
(ClCompile target) ->
  C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\_deps\absl-src\absl\container\internal\compressed_tuple.h(75): fatal error C1001: Internal compiler error. [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\
ortools\glop\ortools_glop.vcxproj]
  cl : command line error D8040: error creating or communicating with child process [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\ortools\glop\ortools_glop.vcxproj]

```

The full command output (big file) is available for viewing as a zipped file with this message.
[CUsersTryerGitDownloadsor-tools-sta.zip](https://github.com/google/or-tools/files/12451426/CUsersTryerGitDownloadsor-tools-sta.zip)

Is there a way to fix these errors?

Thanks.

**Reply by sschnug** [reply]
Please do some own research, e.g. by reading [CMake: Windows: VS Community 2022 can't build or-tools anymore (cl.exe fatal error C1001) #3807](https://github.com/google/or-tools/issues/3807)

**Reply by TryerGit** [reply]
2 errors previously are now down to 1. The error now, after updating to the latest version of Visual Studio 2022 is:

```
"C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL.vcxproj" (default target) (1) ->
(PostBuildEvent target) ->
  C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5):
 error MSB3073: The command "setlocal [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5): e
rror MSB3073: "C:\Program Files\CMake\bin\cmake.exe" -DBUILD_TYPE=Release -P cmake_install.cmake [C:\Users\TryerGit\Downlo
ads\or-tools-stable\or-tools-stable\build\INSTALL.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5): e
rror MSB3073: if %errorlevel% neq 0 goto :cmEnd [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL
.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5): e
rror MSB3073: :cmEnd [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5): e
rror MSB3073: endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone [C:\Users\TryerGit\Downloads\or-tools-stable\or-to
ols-stable\build\INSTALL.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5): e
rror MSB3073: :cmErrorLevel [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5): e
rror MSB3073: exit /b %1 [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5): e
rror MSB3073: :cmDone [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5): e
rror MSB3073: if %errorlevel% neq 0 goto :VCEnd [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL
.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(161,5): e
rror MSB3073: :VCEnd" exited with code 1. [C:\Users\TryerGit\Downloads\or-tools-stable\or-tools-stable\build\INSTALL.vcxpr
oj]

    17 Warning(s)
    1 Error(s)

Time Elapsed 00:23:52.00
```

As you suggested, I will try to do more research to find the underlying cause.

Thanks.

**Comment by TryerGit** [comment]
Ah, had to run the command under x64-Native Tool command prompt as administrator and the problem seems to vanish. :-)
